File: soda/commands/role.py
from typing import Optional

import typer

from soda.context import GlobalContext
from soda.output import render, render_one, print_success, print_error, get_console, output_option
from soda.mock import ROLES

app = typer.Typer(help="Manage roles and permissions.", no_args_is_help=True)


def _get_ctx(ctx: typer.Context) -> GlobalContext:
    return ctx.obj or GlobalContext()


@app.command("list")
def list_cmd(
    ctx: typer.Context,
    output: str = output_option(),
    scope: Optional[str] = typer.Option(None, "--scope", help="Filter by scope: global|dataset"),
):
    """List roles."""
    gctx = _get_ctx(ctx)
    if output != "auto": gctx.output = output
    data = ROLES
    if scope:
        data = [r for r in data if r["scope"] == scope]
    render(data, ["id", "name", "scope", "permissions", "members"], gctx, title="Roles")


@app.command("create")
def create(
    ctx: typer.Context,
    name: str = typer.Option(..., "--name", help="Role name"),
    scope: str = typer.Option(..., "--scope", help="Scope: global|dataset"),
):
    """Create a new role."""
    gctx = _get_ctx(ctx)
    print_success(f"Role [bold]{name}[/bold] created with scope [bold]{scope}[/bold].", gctx)


@app.command("delete")
def delete(
    ctx: typer.Context,
    id: str = typer.Argument(..., help="Role ID"),
):
    """Delete a role."""
    gctx = _get_ctx(ctx)
    print_success(f"Role [bold]{id}[/bold] deleted.", gctx)


@app.command("show")
def show(
    ctx: typer.Context,
    id: str = typer.Argument(..., help="Role ID"),
):
    """Show permissions defined in a role."""
    gctx = _get_ctx(ctx)
    role = next((r for r in ROLES if r["id"] == id), None)
    if not role:
        print_error(f"Role '{id}' not found.", gctx)

    render_one(role, gctx, title=f"Role — {role['name']}")

    # Mock permissions list
    console = get_console(gctx)
    console.print()
    perms_data = [
        {"permission": "datasets:read", "granted": True},
        {"permission": "contracts:write", "granted": True},
        {"permission": "contracts:verify", "granted": True},
        {"permission": "incidents:manage", "granted": False},
        {"permission": "users:manage", "granted": False},
    ]
    render(perms_data, ["permission", "granted"], gctx, title="Permissions")

File: go/cmd/role.go
package cmd

import (
	"fmt"

	"github.com/spf13/cobra"

	"github.com/soda-data-inc/soda-cli/internal/mock"
	"github.com/soda-data-inc/soda-cli/internal/output"
)

var roleCmd = &cobra.Command{
	Use:   "role",
	Short: "Manage roles and their permissions",
}

var roleListCmd = &cobra.Command{
	Use:   "list",
	Short: "List roles",
	RunE: func(cmd *cobra.Command, args []string) error {
		scope, _ := cmd.Flags().GetString("scope")

		rows := mock.Roles
		if scope != "" {
			filtered := []map[string]string{}
			for _, r := range rows {
				if r["scope"] == scope {
					filtered = append(filtered, r)
				}
			}
			rows = filtered
		}

		cols := []string{"id", "name", "scope", "members"}
		output.Render(rows, cols, nil, GCtx)
		return nil
	},
}

var roleCreateCmd = &cobra.Command{
	Use:   "create",
	Short: "Create a new role",
	RunE: func(cmd *cobra.Command, args []string) error {
		name, _ := cmd.Flags().GetString("name")
		scope, _ := cmd.Flags().GetString("scope")

		if name == "" || scope == "" {
			return output.Errorf(2, "--name and --scope are required")
		}

		output.PrintSuccess(fmt.Sprintf("Role '%s' (scope: %s) created.", name, scope), GCtx)
		return nil
	},
}

var roleDeleteCmd = &cobra.Command{
	Use:   "delete <id>",
	Short: "Delete a role",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		output.PrintSuccess(fmt.Sprintf("Role '%s' deleted.", args[0]), GCtx)
		return nil
	},
}

var roleShowCmd = &cobra.Command{
	Use:   "show <id>",
	Short: "Show permissions assigned to a role",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		fmt.Printf("  Permissions for role %s:\n\n", output.Bold.Render(args[0]))
		rows := []map[string]string{
			{"permission": "dataset.read", "scope": "global"},
			{"permission": "contract.verify", "scope": "global"},
			{"permission": "results.read", "scope": "global"},
		}
		cols := []string{"permission", "scope"}
		output.Render(rows, cols, nil, GCtx)
		return nil
	},
}

func init() {
	roleListCmd.Flags().String("scope", "", "Filter by scope: global|dataset")
	roleCreateCmd.Flags().String("name", "", "Role name (required)")
	roleCreateCmd.Flags().String("scope", "", "Role scope: global|dataset (required)")

	roleCmd.AddCommand(roleListCmd, roleCreateCmd, roleDeleteCmd, roleShowCmd)
	rootCmd.AddCommand(roleCmd)
}
